data/modules/TradeShips/Flow.lua

```diff
@@ -0,0 +1,535 @@
+		-- available hyperspace routes lookup table by ship name
+		hyper_routes = hyper_routes,
+		-- station parameters lookup table by it's path
+		port_params = port_params
+	}
+	return true
+end
+
+-- run this function at the start of the game or when jumped to another system
+Flow.spawnInitialShips = function()
+
+	if not Core.params then return nil end
+
+	local hyper_routes = Core.params.hyper_routes
+	local local_routes = Core.params.local_routes
+	local port_params = Core.params.port_params
+
+	-- we want to calculate how many ships of each model are in space, hyperspace
+	-- and stations on average
+	-- it is needed for initial spawn
+
+	-- create tables for all possible ship placement options
+	-- table for ships flying in space
+	local routes_variants = {}
+	-- table for ships docked at station
+	local docked_variants = {}
+	-- amount of ships in space at start
+	local ships_in_space = 0
+	for id, ship_routes in pairs(local_routes) do
+		for _, ship_route in ipairs(ship_routes) do
+			table.insert(
+				routes_variants, {
+					id = id,
+					from = ship_route.from,
+					to = ship_route.to,
+					weight = ship_route.amount
+				})
+			table.insert(
+				docked_variants, {
+					id = id,
+					port = ship_route.to,
+					weight = ship_route.flow
+				})
+			ships_in_space = ships_in_space + ship_route.amount
+		end
+	end
+	local port_summary = {}
+	for k, v in pairs(port_params) do
+		v.port = k
+		table.insert(port_summary, v)
+	end
+	utils.normWeights(routes_variants)
+	utils.normWeights(docked_variants)
+
+	-- amount of ships docked at start
+	local ships_docked = 0
+	for _, params in pairs(port_params) do
+		ships_docked = ships_docked + params.landed
+	end
+
+	-- table for ships in inbound hyperspace cloud
+	local cloud_variants = {}
+	-- amount of ships in cloud at start
+	local ships_in_cloud = 0
+	for id, ship_routes in pairs(hyper_routes) do
+		for _, ship_route in ipairs(ship_routes) do
+			ships_in_cloud = ships_in_cloud + ship_route.ships
+			table.insert(cloud_variants, {
+				ship = id,
+				from = ship_route.from,
+				cloud_duration = ship_route.cloud_duration,
+				weight = ship_route.ships
+			})
+		end
+	end
+	utils.normWeights(cloud_variants)
+
+	-- if we generate ships after jumping into the system, and not when starting a new game,
+	-- count the number of ships that were transferred to this system from the
+	-- source system, and reduce the need for generation by this amount
+	if Core.ships then
+		for ship, trader in pairs(Core.ships) do
+			if trader.status == 'hyperspace' then
+				ships_in_cloud = math.max(0, ships_in_cloud - 1)
+				-- and also assign a random port for this ship, if there are any ports for it at all
+				local routes_for_ship = Core.params.local_routes[ship.shipId]
+				if routes_for_ship then
+					trader.route = utils.chooseNormalized(routes_for_ship)
+				else
+				-- TODO but what if there are no ports for it?
+				end
+			end
+		end
+	else
+		Core.ships = {}
+	end
```

Comment:
Instead of the indentation here, I'd recommend checking for `Core.ships == nil`, creating the table, and moving the loop out of the if statement. That would save one level of indentation, and if the table is newly created will still run no iterations.

```lua
if Core.ships == nil then Core.ships = {} end
for ship, trader in pairs(Core.ships) do
```

Reply:
Done!